tautaunn/binning_algorithms.py

```diff
@@ -1,54 +1,83 @@
+import itertools
+from typing import List, Tuple
+
+
+def yield_requirement(bkgd_values,bkgd_weights, target_val=1e-5):
+    bkgd_weights_cs = np.cumsum(bkgd_weights) 
+    mask = bkgd_weights_cs>target_val
+    if not any(mask):
+        # for this particular shift, the requirement cannot be reached
+        return 1 # ignore this shift for req.
+    else:
+        return bkgd_values[mask][0]
+
+
+def calc_rel_error(weights):
+    return np.sqrt(np.sum(weights**2))/np.sum(weights)
+
+
+def error_requirement(bkgd_values,bkgd_weights, target_val=1.):
+    bkgd_w_cs = np.cumsum(np.abs(bkgd_weights))
+    bkgd_w_cs_2 = np.cumsum(bkgd_weights**2)
+    #neg_mask = bkgd_w_cs>0
+    #bkgd_values = bkgd_values[neg_mask]
+    #bkgd_w_cs = bkgd_w_cs[neg_mask]
+    #bkgd_w_cs_2 = bkgd_w_cs_2[neg_mask]
+    rel_error = np.sqrt(bkgd_w_cs_2)/bkgd_w_cs
+    mask = rel_error<target_val
+    if not any(mask):
+        #print(f"error requirement cannot be reached. returning the min")
+        return 1 # ignore this shift for req. 
+    else:
+        return bkgd_values[mask][0]
+
+
+def sort_vals_and_weights(values, weights):
+    sort_indices = ak.argsort(values, ascending=False)
+    values = values[sort_indices]
+    weights = weights[sort_indices]
+    return values, weights
+
+
+def update_vals_and_weights(vals, weights, edge):
+    mask = vals < edge
+    return vals[mask], weights[mask]
+
+
+def fill_counts(counts, next_edge, hh_shifts, dy_shifts, tt_shifts):
+    hh_mask = hh_shifts["nominal"][0]>=next_edge
+    dy_mask = dy_shifts["nominal"][0]>=next_edge
+    tt_mask = tt_shifts["nominal"][0]>=next_edge
+    counts["HH"][0].append(len(hh_shifts["nominal"][0][hh_mask]))
+    counts["HH"][1].append(np.sum(hh_shifts["nominal"][1][hh_mask]).astype("float64"))
+    counts["HH"][2].append(calc_rel_error(hh_shifts["nominal"][1][hh_mask]).astype("float64"))
+    counts["DY"][0].append(len(dy_shifts["nominal"][0][dy_mask]))
+    counts["DY"][1].append(np.sum(dy_shifts["nominal"][1][dy_mask]).astype("float64"))
+    counts["DY"][2].append(calc_rel_error(dy_shifts["nominal"][1][dy_mask]).astype("float64"))
+    counts["TT"][0].append(len(tt_shifts["nominal"][0][tt_mask]))
+    counts["TT"][1].append(np.sum(tt_shifts["nominal"][1][tt_mask]).astype("float64"))
+    counts["TT"][2].append(calc_rel_error(tt_shifts["nominal"][1][tt_mask]).astype("float64"))
+
+
+def check_yield_requirement(bkgd_values, bkgd_weights, next_edge, target_val=1e-4):
+    mask = bkgd_values >= next_edge
+    return np.sum(bkgd_weights[mask]) > target_val
+
+
+def check_error_requirement(bkgd_values, bkgd_weights, next_edge, target_val=1.):
+    mask = bkgd_values >= next_edge
+    return calc_rel_error(bkgd_weights[mask]) < target_val
+
+
+def get_conditions(dy_vals_weights, tt_vals_weights, next_edge, yield_target=1e-5, error_target=1.):
+    dy_conds = (check_yield_requirement(*dy_vals_weights, next_edge)
+                and check_error_requirement(*dy_vals_weights, next_edge))
```

Comment:
Shouldn't `{yield,error}_target` be forwarded to `check_{yield,error}_requirement`?

Reply:
done